runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/control/DefaultJobBundleFactory.java

```diff
@@ -278,38 +328,51 @@ public RemoteBundle getBundle(
         throws Exception {
       // TODO: Consider having BundleProcessor#newBundle take in an OutputReceiverFactory rather
       // than constructing the receiver map here. Every bundle factory will need this.
-      ImmutableMap.Builder<String, RemoteOutputReceiver<?>> outputReceivers =
-          ImmutableMap.builder();
-      for (Map.Entry<String, Coder> remoteOutputCoder :
-          processBundleDescriptor.getRemoteOutputCoders().entrySet()) {
-        String outputTransform = remoteOutputCoder.getKey();
-        Coder coder = remoteOutputCoder.getValue();
-        String bundleOutputPCollection =
-            Iterables.getOnlyElement(
-                processBundleDescriptor
-                    .getProcessBundleDescriptor()
-                    .getTransformsOrThrow(outputTransform)
-                    .getInputsMap()
-                    .values());
-        FnDataReceiver outputReceiver = outputReceiverFactory.create(bundleOutputPCollection);
-        outputReceivers.put(outputTransform, RemoteOutputReceiver.of(coder, outputReceiver));
-      }
 
-      if (environmentExpirationMillis == 0) {
-        return processor.newBundle(outputReceivers.build(), stateRequestHandler, progressHandler);
+      if (environmentExpirationMillis == 0 && !loadBalanceBundles) {
+        return currentClient.processor.newBundle(
+            getOutputReceivers(currentClient.processBundleDescriptor, outputReceiverFactory)
+                .build(),
+            stateRequestHandler,
+            progressHandler);
       }
 
-      final WrappedSdkHarnessClient client =
-          environmentCaches.get(environmentIndex).getUnchecked(executableStage.getEnvironment());
-      client.ref();
+      final LoadingCache<Environment, WrappedSdkHarnessClient> currentCache;
+      if (loadBalanceBundles) {
+        availableCachesCount.acquire();
+        currentCache = availableCaches.take();
```

Comment:
Why do we need the sempahore and the blocking queue? Shouldn't the queue itself be sufficient? If the queue is not thread-safe, then we could use a simple lock to take from the queue.

Reply:
The semaphore is for fairness, first come first serve.

Reply:
It actually crossed my mind shortly after I wrote this, that only via sempahores we can ensure fair locking. Looks good.